### ISSUE TITLE
the hash of a dna should be based on parsed key/value pairs rather than raw file

### ISSUE THREAD

**Issue by thedavidmeister** [holochain/holochain-proto#654]
Only meaningful changes to a dna should change the hash.

Things that aren't meaningful and so should not change the hash:

- Whether the file is json/yaml/toml
- The order of k/v pairs
- Whitespace in json/toml
- k/v pairs that evaluate to the default value for that key

The same config k/v pairs should always result in connecting to the same network.

This seems mandatory if we ever want the ability to represent dna as anything other than a file, like database rows or in memory.

A secondary benefit of this is that we can then write tests to demonstrate that variants of the same config result in the same network. Then we can release a standard set of tests that can be referenced for anyone wanting to propose an alternative dna implementation.

Byteball has an (unfortunately undocumented) example of something along these lines in their custom JS data serialisation algorithm that is run pre-hashing:

https://github.com/byteball/byteballcore/blob/master/string_utils.js

**Comment by thedavidmeister** [comment]
@pdaoust yes that's exactly what i'm saying, we should hash the logical structure of the config rather than the file it happens to be stored in

"normalising" to a standard format that all other variants can be converted to would work fine, that's basically what the byteball example does (they just invented their own serialisation format)

- sorting on key fixes order ✔
- converting to json _then_ minifying fixes whitespace (yaml has meaningful whitespace) ✔
- converting to json removes inconsistencies caused by file format ✔

the only other thing to do is merge any default values into the json...

so in order:

1. convert to json
2. merge in default values
3. sort by key
4. minify
5. hash

**Comment by pdaoust** [comment]
@thedavidmeister Ahhhhhhhh... I understand your context now! The reason you're restricting the conversation to _just_ the config part of the DNA is because you're talking about spinning up ad-hoc versions of the same app (same zomes) with a different hash -- for the 'private chats' or 'circles' pattern. Yeah, I appreciate that a deterministic structuring of the `dna.json` (or a synthesis of that file and config passed in via an agent) would be crucial, since you're not looking at a file you can read or lint anymore.

**Comment by thedavidmeister** [comment]
yes, it's _just_ the config, because managing network config and the code being executed on the network are pretty different things, pragmatically speaking

private chat is the most obvious use case yes

more generally, you want this any time you want to _generate_ dna.json files with different systems/methods and have them reliably "talk to each other".

relevant other discussion:

https://github.com/holochain/holochain-proto/issues/662
https://github.com/holochain/holochain-proto/issues/656
